### PR TITLE
Merchants update products

### PR DESCRIPTION
**Pull request by msfinnan**
Merchant can edit and update their own products
-only logged in merchant can do this, link only shows on their product pages (to see this you will need to create a product if you have not already)
-there is some code in the controllers to take care of how money shows in the form. This is not ideal, but it works. I will investigate other ways to deal with this. It might be a gem we install. 
-one flash message is not working, there is a note in the code and the test. Not urgent bc the form also validates for all fields. 
-all tests are green, let me know if there are any tests I'm missing.

*if you download this branch and it looks good online, I'm ready for it to be merged. I can work on the money display thing in another branch*

### CODE REVIEW

File: test/controllers/products_controller_test.rb [hunk 1]
```diff
@@ -79,32 +68,129 @@
...
-    it "renders bad_request and does not update the DB for bogus data" do
-      merchant = perform_login
-      bad_prod_name = { product: { name: nil, price: 1000 } }
+        get edit_product_path(-1)
 
-      expect {
-        post products_path, params: bad_prod_name
-      }.wont_change "Product.count"
+        must_respond_with :redirect
+      end
```
Comment (qqdipps) [inline test/controllers/products_controller_test.rb:115]: where does it redirect to?

File: test/controllers/products_controller_test.rb [hunk 2]
```diff
@@ -79,32 +68,129 @@
...
+      it "redirects user" do
+        merchant = merchants(:merchant_1)
+        # do not do login
+        get edit_product_path(merchant.products.first.id)
 
-      must_respond_with :bad_request
-      expect(flash[:error]).must_include "Could not add new product"
-      expect(flash[:name]).must_include "can't be blank"
+        # must_respond_with :redirect
+        expect(flash[:error]).must_equal "You must be logged in to edit a product"
```
Comment (qqdipps) [inline test/controllers/products_controller_test.rb:124]: is this meant to be commented out?

File: test/controllers/products_controller_test.rb [hunk 3]
```diff
@@ -79,32 +68,129 @@
...
-      must_respond_with :bad_request
-      expect(flash[:error]).must_equal "You must be logged in to create a new product"
+        expect {
+          patch product_path(product_to_update.id), params: test_input
+        }.wont_change "Product.count"
+
+        must_respond_with :bad_request
+        # expect(flash[:name]).must_include "can't be blank" # not working
+      end
+    end
```
Comment (qqdipps) [inline test/controllers/products_controller_test.rb:190]: this works fine. I uncommented and changed comment to working so you can still find it.

File: app/controllers/products_controller.rb [hunk 4]
```diff
@@ -49,9 +47,41 @@ def show
...
+      @product.price = product_params[:price].to_f * 100.0
+      is_successful = @product.save
+
+      if is_successful
+        flash[:success] = "Product updated successfully"
+        redirect_to product_path(@product.id)
+      else
+        @product.errors.messages.each do |label, message|
+          flash.now[label.to_sym] = message #this flash msg is working
+        end
```
Comment (qqdipps) [inline app/controllers/products_controller.rb:78]: works fine. uncommented test. test passes. =)

File: test/controllers/products_controller_test.rb [hunk 5]
```diff
@@ -79,32 +68,129 @@
...
+        }
+
+        expect {
+          patch product_path(product_to_update.id), params: test_input
+        }.wont_change "Product.count"
+
+        must_respond_with :redirect
+        product_to_update.reload
+        expect(product_to_update.name).must_equal test_input[:product][:name]
+        expect(product_to_update.merchant_id).must_equal test_input[:product][:merchant_id]
```
Comment (qqdipps) [inline test/controllers/products_controller_test.rb:163]: `must_redirect_to  some_path`

File: test/controllers/products_controller_test.rb (hunk 2 above)
Comment (qqdipps) [reply test/controllers/products_controller_test.rb:124]: also if it does redirect, where does it redirect to?

**Review by qqdipps — APPROVED** [review]
pulled down code, messed with features, no bugs noted. lgtm, just checkout few notes on testing and stuffs.

File: test/controllers/products_controller_test.rb (hunk 2 above)
Comment (msfinnan) [reply test/controllers/products_controller_test.rb:124]: not meant to be commented out, added a redirect to products_path :)

File: test/controllers/products_controller_test.rb (hunk 3 above)
Comment (msfinnan) [reply test/controllers/products_controller_test.rb:190]: thx! took out comment

File: app/controllers/products_controller.rb (hunk 4 above)
Comment (msfinnan) [reply app/controllers/products_controller.rb:78]: cool!

File: test/controllers/products_controller_test.rb (hunk 5 above)
Comment (msfinnan) [reply test/controllers/products_controller_test.rb:163]: done!

File: test/controllers/products_controller_test.rb (hunk 1 above)
Comment (msfinnan) [reply test/controllers/products_controller_test.rb:115]: fixed!

**Pull request merged**